src/navier_stokes/VCINSStaggeredHierarchyIntegrator.cpp

```diff
@@ -0,0 +1,3295 @@
+                                 << " requires num_cycles > 1.\n"
+                                 << "  at current time step, num_cycles = "
+                                 << d_current_num_cycles
+                                 << "\n");
+    }
+
+    // Allocate the scratch and new data.
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        Pointer<PatchLevel<NDIM> > level = d_hierarchy->getPatchLevel(ln);
+        level->allocatePatchData(d_scratch_data, current_time);
+        level->allocatePatchData(d_new_data, new_time);
+        level->allocatePatchData(d_velocity_C_idx, current_time);
+        level->allocatePatchData(d_velocity_rhs_C_idx, current_time);
+        level->allocatePatchData(d_velocity_D_idx, current_time);
+        level->allocatePatchData(d_velocity_rhs_D_idx, current_time);
+        level->allocatePatchData(d_pressure_D_idx, current_time);
+        level->allocatePatchData(d_pressure_rhs_D_idx, current_time);
+        level->allocatePatchData(d_temp_sc_idx, current_time);
+        level->allocatePatchData(d_temp_cc_idx, current_time);
+        level->allocatePatchData(d_rho_interp_idx, current_time);
+        level->allocatePatchData(d_mu_interp_idx, current_time);
+    }
+
+    // Compute variable mass density.
+    if (d_rho_var)
+    {
+        if (d_rho_fcn)
+        {
+            d_rho_fcn->setDataOnPatchHierarchy(d_rho_current_idx, d_rho_var, d_hierarchy, current_time);
+            d_rho_fcn->setDataOnPatchHierarchy(d_rho_new_idx, d_rho_var, d_hierarchy, new_time);
+        }
+        else
+        {
+            d_hier_cc_data_ops->copyData(d_rho_new_idx, d_rho_current_idx);
+        }
+        d_hier_cc_data_ops->linearSum(d_rho_scratch_idx, 0.5, d_rho_current_idx, 0.5, d_rho_new_idx);
+        d_rho_bdry_bc_fill_op->fillData(half_time);
+
+        for (int level_num = coarsest_ln; level_num <= finest_ln; ++level_num)
+        {
+            Pointer<PatchLevel<NDIM> > level = d_hierarchy->getPatchLevel(level_num);
+            for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+            {
+                Pointer<Patch<NDIM> > patch = level->getPatch(p());
+                Pointer<CellData<NDIM, double> > temp_data = patch->getPatchData(d_temp_cc_idx);
+                Pointer<CellData<NDIM, double> > rho_data = patch->getPatchData(d_rho_scratch_idx);
+                for (int d = 0; d < NDIM; ++d) temp_data->copyDepth(d, (*rho_data), 0);
+            }
+        }
+        // Interpolate onto side centers
+        d_hier_math_ops->harmonic_interp(
+            d_rho_interp_idx, d_rho_interp_var, false, d_temp_cc_idx, d_temp_cc_var, d_no_fill_op, half_time);
+    }
+
+    // Compute variable viscosity.
+    if (d_mu_var)
+    {
+        if (d_mu_fcn)
+        {
+            d_mu_fcn->setDataOnPatchHierarchy(d_mu_current_idx, d_mu_var, d_hierarchy, current_time);
+            d_mu_fcn->setDataOnPatchHierarchy(d_mu_new_idx, d_mu_var, d_hierarchy, new_time);
+        }
```

Comment:
Similarly, for mu. 


There should however be callback functions to reset rho/mu here according to the physics of the problem.